File: go/spore-core/sensor/sensor_fixture_replay_test.go
package sensor

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
	"testing"
)

// fixtureSensor / fixtureEvent / fixtureExpected mirror the JSON shape in
// fixtures/sensor_chain/*.json. The structures must match the Rust reference
// (`rust/crates/spore-core/src/sensor.rs::FixtureCase`) so the shared
// fixture deserializes here without modification.

type fixtureSensor struct {
	ID         string                 `json:"id"`
	Kind       SensorKind             `json:"kind"`
	Triggers   []SensorTrigger        `json:"triggers"`
	Outcome    SensorOutcome          `json:"outcome"`
	Thresholds SensorSignalThresholds `json:"thresholds"`
}

type fixtureEvent struct {
	Trigger   SensorTrigger `json:"trigger"`
	SessionID string        `json:"session_id"`
}

type fixtureExpected struct {
	NeverFired   []string `json:"never_fired"`
	AlwaysFiring []string `json:"always_firing"`
}

type fixtureCase struct {
	Sensors     []fixtureSensor `json:"sensors"`
	Events      []fixtureEvent  `json:"events"`
	MinSessions uint32          `json:"min_sessions"`
	Expected    fixtureExpected `json:"expected"`
}

// TestSensorChainFixtureReplay replays the shared signal_quality_basic.json
// fixture and asserts the same NeverFired / AlwaysFiring sets the Rust
// reference produces.
func TestSensorChainFixtureReplay(t *testing.T) {
	wd, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	// go/spore-core/sensor → ../../../fixtures/sensor_chain/...
	path := filepath.Join(wd, "..", "..", "..", "fixtures", "sensor_chain", "signal_quality_basic.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fc fixtureCase
	if err := json.Unmarshal(raw, &fc); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}

	chain := NewStandardSensorChain()
	for _, s := range fc.Sensors {
		cfg := SensorConfig{
			ID:                 SensorID(s.ID),
			Name:               s.ID,
			Kind:               s.Kind,
			Triggers:           s.Triggers,
			LowSignalThreshold: s.Thresholds,
		}
		if err := chain.Register(context.Background(), &stubSensor{cfg: cfg, outcome: s.Outcome}); err != nil {
			t.Fatalf("register %s: %v", s.ID, err)
		}
	}
	for _, ev := range fc.Events {
		chain.Fire(context.Background(), ev.Trigger, input(ev.SessionID))
	}
	flags := chain.SignalQualityReport(context.Background(), fc.MinSessions)

	gotNever := []string{}
	gotAlways := []string{}
	for _, f := range flags {
		switch f.Kind {
		case FlagKindNeverFired:
			gotNever = append(gotNever, string(f.SensorID))
		case FlagKindAlwaysFiring:
			gotAlways = append(gotAlways, string(f.SensorID))
		}
	}
	sort.Strings(gotNever)
	sort.Strings(gotAlways)
	wantNever := append([]string(nil), fc.Expected.NeverFired...)
	wantAlways := append([]string(nil), fc.Expected.AlwaysFiring...)
	sort.Strings(wantNever)
	sort.Strings(wantAlways)

	if !equalStringSlices(gotNever, wantNever) {
		t.Errorf("never_fired mismatch: got %v want %v", gotNever, wantNever)
	}
	if !equalStringSlices(gotAlways, wantAlways) {
		t.Errorf("always_firing mismatch: got %v want %v", gotAlways, wantAlways)
	}
}

func equalStringSlices(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: typescript/packages/core/tests/sensor-fixture-replay.test.ts
/**
 * Fixture-replay tests for the canonical SensorChain (spore-core issue #10).
 *
 * Loads `fixtures/sensor_chain/signal_quality_basic.json` and asserts the
 * NeverFired / AlwaysFiring sets exactly match the Rust, Python, and Go
 * suites.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { sensor, SessionId } from "../src/index.js";
import { emptySessionState } from "../src/harness/types.js";

const { StandardSensorChain, SensorId, newSensorInput } = sensor;

interface FixtureSensor {
  id: string;
  kind: sensor.SensorKind;
  triggers: sensor.SensorTrigger[];
  outcome: sensor.SensorOutcome;
  thresholds: sensor.SensorSignalThresholds;
}
interface FixtureEvent {
  trigger: sensor.SensorTrigger;
  session_id: string;
}
interface FixtureCase {
  name: string;
  description?: string;
  sensors: FixtureSensor[];
  events: FixtureEvent[];
  min_sessions: number;
  expected: { never_fired: string[]; always_firing: string[] };
}

const here = dirname(fileURLToPath(import.meta.url));
const fixturePath = resolve(here, "../../../../fixtures/sensor_chain/signal_quality_basic.json");

class StubSensor implements sensor.Sensor {
  constructor(
    private readonly cfg: sensor.SensorConfig,
    private readonly outcome: sensor.SensorOutcome,
  ) {}
  async evaluate(): Promise<sensor.SensorResult> {
    return {
      sensor_id: new SensorId(this.cfg.id.value),
      outcome: this.outcome,
      observation: this.outcome === "warn" ? "warn-obs" : null,
      detail: this.outcome,
      fired_at: new sensor.Timestamp("2026-05-16T00:00:00Z"),
    };
  }
  config(): sensor.SensorConfig {
    return this.cfg;
  }
}

describe("SensorChain fixture replay", () => {
  it("signal_quality_basic produces expected flags", async () => {
    const raw = readFileSync(fixturePath, "utf8");
    const fix = JSON.parse(raw) as FixtureCase;

    const chain = new StandardSensorChain();
    for (const s of fix.sensors) {
      const cfg: sensor.SensorConfig = {
        id: new SensorId(s.id),
        name: s.id,
        kind: s.kind,
        triggers: s.triggers,
        run_every_n_turns: null,
        run_on_phases: null,
        low_signal_threshold: s.thresholds,
      };
      chain.register(new StubSensor(cfg, s.outcome));
    }
    for (const ev of fix.events) {
      await chain.fire(
        ev.trigger,
        newSensorInput(new SessionId(ev.session_id), emptySessionState()),
      );
    }
    const flags = await chain.signalQualityReport(fix.min_sessions);

    const gotNever = new Set<string>();
    const gotAlways = new Set<string>();
    for (const f of flags) {
      if (f.kind === "never_fired") gotNever.add(f.sensor_id.value);
      else gotAlways.add(f.sensor_id.value);
    }
    expect([...gotNever].sort()).toEqual([...fix.expected.never_fired].sort());
    expect([...gotAlways].sort()).toEqual([...fix.expected.always_firing].sort());
  });
});
